base_layer/core/src/chain_storage/blockchain_database.rs

```diff
@@ -352,6 +352,65 @@ where B: BlockchainBackend
             self.clear_all_reorgs()?;
         }
 
+        self.rebuild_payref_indexes_task()?;
+
+        Ok(())
+    }
+
+    fn rebuild_payref_indexes_task(&self) -> Result<(), ChainStorageError> {
+        let initial_status = {
+            let db = self.db_read_access()?;
+            db.fetch_payref_rebuild_status()?
+        };
+        if !initial_status.is_rebuilt {
+            let chain_metadata = {
+                let db = self.db_read_access()?;
+                db.fetch_chain_metadata()?
+            };
+            let db_rw_lock = self.db.clone();
+
+            tokio::spawn(async move {
+                let start_height = initial_status.last_rebuild_height.unwrap_or_default();
+                let mut last_status = initial_status.clone();
+                debug!(
+                    target: LOG_TARGET,
+                    "[PayRef] Starting index rebuilding for heights {} to {}",
+                    start_height, chain_metadata.best_block_height()
+                );
+                if let Ok(write_txn) = db_rw_lock
+                    .write()
+                    .map_err(|_e| ChainStorageError::AccessError("Write lock on blockchain backend failed".into()))
+                {
+                    write_txn.set_stats_total_height(chain_metadata.best_block_height());
+                }
+                for height in start_height..=chain_metadata.best_block_height() {
+                    match process_payref_for_height(
+                        db_rw_lock.clone(),
+                        height,
+                        chain_metadata.clone(),
+                        height == chain_metadata.best_block_height(),
+                    ) {
+                        Ok(current_status) => {
+                            last_status = current_status;
+                            if let Ok(write_txn) = db_rw_lock.write().map_err(|_e| {
+                                ChainStorageError::AccessError("Write lock on blockchain backend failed".into())
+                            }) {
+                                write_txn.update_stats_progress(height);
+                            }
```

Comment:
```suggestion

```
This should not be done on a new write txn, should be the same one used to update the payrefs for the height.